tests/test_vrf.py

```diff
@@ -0,0 +1,1395 @@
+        vrf_member_port_indices[vrf] = []
+
+        for intf in intfs:
+            vrf_intf_member_port_indices[vrf][intf] = natsorted(
+                    [ cfg_facts['config_port_indices'][m.split('|')[1]] for m in filter(lambda m: intf in m, member) ]
+                )
+            vrf_member_port_indices[vrf].extend(vrf_intf_member_port_indices[vrf][intf])
+
+        vrf_member_port_indices[vrf] = natsorted(vrf_member_port_indices[vrf])
+
+    return vrf_intf_member_port_indices, vrf_member_port_indices
+
+def ex_ptf_runner(ptf_runner, exc_queue, **kwargs):
+    '''
+    With this simple warpper function, we could use a Queue to store the
+    exception infos and check it later in main thread.
+
+    Example:
+        refer to test 'test_vrf_swss_warm_reboot'
+    '''
+    try:
+        ptf_runner(**kwargs)
+    except Exception:
+        exc_queue.put(sys.exc_info())
+
+def finalize_warmboot(duthost, comp_list=None, retry=30, interval=5):
+    '''
+    Check if componets finish warmboot(reconciled).
+    '''
+    DEFAULT_COMPONENT_LIST = ['orchagent', 'neighsyncd']
+    EXP_STATE = 'reconciled'
+
+    comp_list = comp_list or DEFAULT_COMPONENT_LIST
+
+    # wait up to $retry * $interval secs
+    for _ in range(retry):
+        for comp in comp_list:
+            state =  duthost.shell('/usr/bin/redis-cli -n 6 hget "WARM_RESTART_TABLE|{}" state'.format(comp), module_ignore_errors=True)['stdout']
+            print "{} : {}".format(comp, state)
+            if EXP_STATE == state:
+                comp_list.remove(comp)
+        if len(comp_list) == 0:
+            break
+        time.sleep(interval)
+
+    return  comp_list
+
+def setup_vrf_cfg(duthost, cfg_facts):
+    '''
+    setup vrf configuration on dut before test suite
+    '''
+
+    # FIXME
+    # For vrf testing, we should create a new vrf topology
+    # might named to be 't0-vrf', deploy with minigraph templates.
+    #
+    # But currently vrf related schema does not properly define in minigraph.
+    # So we generate and deploy vrf basic configuration with a vrf jinja2 template,
+    # later should move to minigraph or a better way.
+
+    from copy import deepcopy
+    cfg_t0 = deepcopy(cfg_facts)
+
+    cfg_t0.pop('config_port_indices', None)
+
+    # get members from Vlan1000, and move half of them to Vlan2000 in vrf basic cfg
+    ports = get_vlan_members('Vlan1000', cfg_facts)
+
+    vlan_ports = {'Vlan1000': ports[:len(ports)/2],
+                  'Vlan2000': ports[len(ports)/2:]}
+
+    extra_vars = {'cfg_t0': cfg_t0,
+                  'vlan_ports': vlan_ports}
+
+    duthost.host.options['variable_manager'].extra_vars = extra_vars
+
+    #backup config_db.json
+    duthost.shell("mv /etc/sonic/config_db.json /etc/sonic/config_db.json.bak")
+
+    duthost.template(src="vrf/vrf_config_db.j2", dest="/tmp/config_db_vrf.json")
+    duthost.shell("cp /tmp/config_db_vrf.json /etc/sonic/config_db.json")
+
+    # FIXME use a better way to load config
```

Comment:
Indeed this is not a good way to load new config.  If my understanding is correct, the new generated config_DB lack of METADATA, PORT, mgmt interface and so on, I am not sure whether it can properly boot up the switch, maybe work on your platform, but could have issues on others. Since there is CLI support for VRF, why not using it? or sonic-cfggen should also work.

Reply:
The new generated config_db include all sections of what you mentioned. Please refer to https://github.com/dawnbeauty/sonic-mgmt/blob/98beced8a86f32340a46c277e2bab1c55c34cbb7/tests/vrf/vrf_config_db.j2#L98

Currently the VRF cli could not setup configuration of BGP_NEIGHBOR and BGP_PEER_RANGE with vrf, so i prefer a jinja2 template instead of cli commands.

Reply:
ok for me.